src/librustc_mir/borrow_check.rs

```diff
@@ -846,6 +850,199 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
+                            ty::TyRef(_, tnm) => {
+                                match tnm.mutbl {
+                                    // Shared borrowed data is never mutable
+                                    hir::MutImmutable => Err(lvalue),
+                                    // Mutably borrowed data is mutable, but only if we have a
+                                    // unique path to the `&mut`
+                                    hir::MutMutable => self.can_host_mutable_ref(&proj.base),
+                                }
+                            },
+                            ty::TyRawPtr(tnm) => {
+                                match tnm.mutbl {
+                                    // `*const` raw pointers are not mutable
+                                    hir::MutImmutable => Err(lvalue),
+                                    // `*mut` raw pointers are always mutable, regardless of context
+                                    // The users have to check by themselve.
+                                    hir::MutMutable => Ok(()),
+                                }
+                            },
+                            // All other projections are owned by their base path, so mutable if
+                            // base path is mutable
+                            _ => self.is_mutable(&proj.base)
+                        }
+                    },
+                    // All other projections are owned by their base path, so mutable if
+                    // base path is mutable
+                    _ => self.is_mutable(&proj.base)
+                }
+            }
+        }
+    }
+
+    /// This lvalue owns an `&mut`, can it be used to write mutably?
+    ///
+    /// The basic idea is to enforce a unique path.
+    fn can_host_mutable_ref<'d>(&self, lvalue: &'d Lvalue<'tcx>) -> Result<(), &'d Lvalue<'tcx>> {
+        match *lvalue {
+            Lvalue::Local(..) => {
+                // Local variables are unique
+                Ok(())
+            },
+            Lvalue::Static(..) => {
+                // Static variables are not
+                Err(lvalue)
+            },
+            Lvalue::Projection(ref proj) => {
+                match proj.elem {
+                    ProjectionElem::Deref => {
+                        let base_ty = proj.base.ty(self.mir, self.tcx).to_ty(self.tcx);
+
+                        // `Box<T>` referent is unique if box is a unique spot
+                        if base_ty.is_box() {
+                            return self.can_host_mutable_ref(&proj.base);
+                        }
+
+                        // Otherwise we check the kind of deref to decide
+                        match base_ty.sty {
+                            ty::TyRef(_, tnm) => {
+                                match tnm.mutbl {
+                                    // lvalue represent an aliased location
+                                    hir::MutImmutable => Err(lvalue),
```

Comment:
Actually, it appears we *do* allow the use of `&mut` there. [This example compiles:](https://play.rust-lang.org/?gist=e8e300fa7063f51b1b2f449974c30ae8&version=stable)

```
fn foo(x: *const &mut u32) {
  unsafe {
    **x += 1;
  }
}

fn main() { }
```

Reply:
So I was wrong in my writeup, we should just treat `*const` the same as `*mut`.

Reply:
Er, I meant this for the raw pointer cases below, sorry =) comment on wrong line